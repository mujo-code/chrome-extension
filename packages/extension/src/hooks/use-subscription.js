import { Payment, Functional } from '@mujo/utils'
import React, {
  useState,
  useEffect,
  useCallback,
  useContext,
} from 'react'
import { ACTIVE_PRODUCT } from '../constants'

const { compose, first } = Functional

export const activeProducts = products =>
  products.filter(product => product.state === ACTIVE_PRODUCT)

export const userFactory = (products = []) => ({
  isSubscribed: !!activeProducts(products).length,
  products,
  has: product => products.includes(product),
})

export const hydrate = async ({ setProducts, setUser }) => {
  const hydrateUser = compose(setUser, userFactory)
  hydrateUser(await Payment.getPurchases())
  setProducts(await Payment.getProducts())
}

const context = React.createContext({ user: userFactory([]) })
export const SubscriptionProvider = props => {
  const [isInitialized, setIsInitialized] = useState(false)
  const [products, setProducts] = useState([])
  const [user, setUser] = useState(userFactory())
  const [purchaseError, setPurchaseError] = useState(null)

  const getProduct = useCallback(
    sku => first(products.filter(product => product.sku === sku)),
    [products]
  )

  const buy = useCallback(
    async sku => {
      setPurchaseError(null) // reset error
      try {
        await Payment.buy(sku)
      } catch (e) {
        setPurchaseError(e)
        return
      }
      // eager update user
      setUser({
        ...user,
        products: [getProduct(sku)],
        isSubscribed: true,
      })
    },
    [getProduct, user]
  )

  useEffect(() => {
    if (!isInitialized) {
      setIsInitialized(true)
      hydrate({ setProducts, setUser })
    }
  }, [setProducts, setUser, isInitialized, setIsInitialized])

  const { Provider } = context

  const value = {
    buy,
    getProduct,
    isInitialized,
    products,
    purchaseError,
    user,
  }
  return <Provider {...props} value={value} />
}

export const useSubscription = () => useContext(context)
