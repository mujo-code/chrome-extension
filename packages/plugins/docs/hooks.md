# Hooks

Hooks will allow plugins to get access to many of the applications internal API's. Things like setting up some keys in the database or event listening to the internal alarm system. Some of these hooks can run on all of the scripts and will look like from your application that the hooks are global working between the scripts.

## Universal Hooks

### useStorage

This hook should allow the plugin to access data inside the database. It should allow for access to the database between all the scripts.

```javascript
const [value, setValue] = useStorage('KEY', defaultValue)
```

The data is all stored locally inside the browser inside indexDB.

## Background Hooks

### useHeartBeat

This hook will allow plugins to listen to the heartbeat of the application. Its is a 15-minute tick that the background process has. It is also based on alarms so the timing is not precise at all and sometimes can be deferred by the browser. `useHeartBeat` should be used for triggers based on usage. This allows the plugin to interject at times of activity. It gets a single argument to the effect function which is if the user is currently active.

```javascript
useHeartBeat(isActive => {
  if (isActive && otherCondition) return doSomething()
}) // TODO maybe have dependencies here for cache busting
```

### useMessage

This hook is will allow plugins to receive messages from NTP and content to background scripts. This interface is unified between all scripts for ease of use.

```javascript
useMessage((type, body, response) => {
  // do something with message
  // TODO: figure out interface for sending out messages
  // eg. responses, and broadcasting on background or sending a message
  // on client scripts.
})
```
