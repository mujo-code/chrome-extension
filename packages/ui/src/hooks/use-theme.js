import React, { useContext } from 'react'

const DEFAULT_COLOR_THEME = 'dark'

const colors = {
  dark: {
    foreground: 'light',
    foregroundSecondary: 'white',
    background: 'dark',
    backgroundSecondary: 'black',
    highlight: 'highlight',
    buttonStyle: 'primary',
  },
  light: {
    foreground: 'dark',
    foregroundSecondary: 'black',
    background: 'light',
    backgroundSecondary: 'white',
    highlight: 'highlight',
    buttonStyle: 'tertiary',
  },
}

export const ColorTheme = React.createContext(DEFAULT_COLOR_THEME)
export const ColorThemeProvider = ColorTheme.Provider
export const ColorThemeConsumer = ColorTheme.Consumer

export const useTheme = () => {
  const theme = useContext(ColorTheme)
  return colors[theme] || colors.dark
}
