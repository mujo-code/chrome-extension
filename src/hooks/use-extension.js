import { useEffect, useState } from 'react'
import {
  ALARM_KEY,
  TOP_SITES_KEY,
  TOP_SITES_USAGE_KEY,
  PAGE_VIEWS_KEY,
  SITE_TIME_KEY,
  SHOW_TOP_SITES_KEY,
  NEW_TAB_CONNECTION,
  CLEAR_ALARM,
  SET_ALARM,
  BREAK_TIMERS_KEY,
} from '../constants'
import { message, topSites as topSitesApi } from '../lib/extension'
import { first } from '../lib/functional'
import { set, create } from '../lib/util'
import { useStorage } from './use-storage'

export const onStorageChange = ({ setSiteTimes }) => e => {
  // TODO: make this refrersh all keys, need better way to access refresh
  switch (e.key) {
    case SITE_TIME_KEY:
      // refresh will update value from localStorage
      setSiteTimes(null, { refresh: true })
      break
    default:
  }
}

export const useExtension = () => {
  const [appReady, setAppReady] = useState(false)
  const [selectedSegment, setSelectedSegment] = useState(null)
  const [alarmEnabled, setAlarmEnabled] = useStorage(ALARM_KEY)
  const [topSites, setTopSites] = useStorage(TOP_SITES_KEY)
  const [siteTimes, setSiteTimes] = useStorage(SITE_TIME_KEY)
  const [topSitesUsage, setTopSitesUsage] = useStorage(
    TOP_SITES_USAGE_KEY
  )
  const [pageViews, updatePageViews, { pending }] = useStorage(
    PAGE_VIEWS_KEY
  )
  const [showTopSites, updateShowTopSites] = useStorage(
    SHOW_TOP_SITES_KEY
  )
  const [breakTimers, updateBreakTimers] = useStorage(
    BREAK_TIMERS_KEY
  )

  // mount hook
  useEffect(
    () => {
      if (!pending && !appReady) {
        updatePageViews(pageViews + 1)
        topSitesApi.get(setTopSites)
        window.addEventListener(
          'storage',
          onStorageChange({ setSiteTimes })
        )
        message(NEW_TAB_CONNECTION)
        setAppReady(true)
      }
    },
    [
      pageViews,
      pending,
      setSiteTimes,
      setTopSites,
      updatePageViews,
      appReady,
    ] /* NOTE: stops useEffect from continuing firing */
  )

  const setAlarmEnabledProxy = () => {
    const enabled = !alarmEnabled
    setAlarmEnabled(enabled)
    if (enabled) {
      message(SET_ALARM)
    } else {
      message(CLEAR_ALARM)
    }
  }

  const setBreakTimer = (url, time, enabled) => {
    const lastBreakTimer = breakTimers[url] || {}
    const nextBreakTimer = Object.assign({}, lastBreakTimer, {
      url,
      time,
      enabled,
    })
    const nextBreakTimers = create(breakTimers, url, nextBreakTimer)
    updateBreakTimers(nextBreakTimers)
  }

  const updateSitesUsed = site => {
    setTopSitesUsage([...topSitesUsage, site])
  }

  const resetUsage = () => {
    updatePageViews(0)
    setTopSitesUsage([])
  }

  const mappedTopSites = topSites.map(topSite => {
    const { url } = topSite
    const isUsed = topSitesUsage.some(site => site.url === url)
    return Object.assign({ isUsed }, topSite)
  })

  const siteTimesAndTimers = Object.keys(siteTimes).reduce(
    (accum, url) => {
      set(accum, url, accum[url] || {})
      set(accum[url], 'time', siteTimes[url])
      set(accum[url], 'breakTimer', breakTimers[url] || {})
      return accum
    },
    {}
  )

  if (selectedSegment) {
    const { urls } = selectedSegment
    if (urls.length === 1) {
      const originalURL = first(selectedSegment.urls)
      const { breakTimer, time } = siteTimesAndTimers[originalURL]
      set(selectedSegment, 'data', {
        breakTimer,
        time,
        originalURL,
      })
    }
  }

  return [
    {
      topSites: mappedTopSites,
      alarmEnabled,
      pageViews,
      showTopSites,
      siteTimes,
      siteTimesAndTimers,
      appReady,
      breakTimers,
      selectedSegment,
    },
    {
      setAlarmEnabled: setAlarmEnabledProxy,
      setTopSites,
      updateSitesUsed,
      resetUsage,
      updateShowTopSites,
      setBreakTimer,
      setSelectedSegment,
    },
  ]
}
