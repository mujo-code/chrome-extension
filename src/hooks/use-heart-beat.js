import { useEffect } from 'react'
import { alarms } from '../background/alarm'
import { HEARTBEAT } from '../constants'

export const useHeartBeat = fn => {
  useEffect(() => {
    alarms.on(HEARTBEAT, fn)
    return () => alarms.off(HEARTBEAT, fn)
  }, [fn])
}
