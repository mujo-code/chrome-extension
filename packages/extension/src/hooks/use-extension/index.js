import { IngressProvider } from '@mujo/ingress'
import { Extension } from '@mujo/utils'
import React, {
  useEffect,
  useState,
  useContext,
  useCallback,
} from 'react'
import {
  NEW_TAB_CONNECTION,
  RESET_USAGE,
  APP_READY_KEY,
  CURRENT_TAB_KEY,
} from '../../constants'
import { useSubscription } from '../use-subscription'
import { decorateSelectedSegment, mapTopSites } from './transforms'
import { useBreaktimerCallback } from './use-breaktimer-callback'
import { useDeeplink } from './use-deeplink'
import { useModel } from './use-model'
import { useSettings } from './use-settings'
import { useTabs } from './use-tabs'
import { useEndScreen } from './use-end-screen'
import { useTopsitesAPI } from './use-topsites-api'

const { message } = Extension

const context = React.createContext()
const { Provider } = context

export const ExtensionProvider = props => {
  // state
  const { shouldRegisterApp } = props
  const [appReady, setAppReady] = useState(false)
  const [playerIsOpen, setPlayerIsOpen] = useState(false)
  const [selectedSegment, setSelectedSegment] = useState(null)
  const [upsellModal, setUpsellModal] = useState(null)
  const tabInterface = useTabs(CURRENT_TAB_KEY)
  const {
    alarmEnabled,
    setAlarmEnabled,
    topSites,
    setTopSites,
    siteTimes,
    setSiteTimes,
    topSitesUsage,
    setTopSitesUsage,
    pageViews,
    updatePageViews,
    pending,
    showTopSites,
    updateShowTopSites,
    breakTimers,
    updateBreakTimers,
    activityNumber,
    breathAmount,
    setBreathAmount,
  } = useModel()
  const { user } = useSubscription()
  const endScreenInterface = useEndScreen()

  // effects
  useDeeplink({
    appReady,
    setPlayerIsOpen,
    setSelectedSegment,
  })
  useTopsitesAPI({ setTopSites })
  useEffect(() => {
    if (!pending && !appReady && shouldRegisterApp) {
      updatePageViews(pageViews + 1)
      message(NEW_TAB_CONNECTION)
      setAppReady(true)
      if (typeof window[APP_READY_KEY] === 'function') {
        window[APP_READY_KEY]()
      }
      window[APP_READY_KEY] = true
    }
  }, [
    pageViews,
    pending,
    setSiteTimes,
    setTopSites,
    updatePageViews,
    appReady,
    shouldRegisterApp,
  ])

  // callbacks
  const setAlarmEnabledProxy = useCallback(() => {
    const enabled = !alarmEnabled
    setAlarmEnabled(enabled)
  }, [alarmEnabled, setAlarmEnabled])
  const { setBreakTimer } = useBreaktimerCallback({
    breakTimers,
    setUpsellModal,
    updateBreakTimers,
    user,
  })
  const updateSitesUsed = useCallback(
    site => {
      setTopSitesUsage([...topSitesUsage, site])
    },
    [setTopSitesUsage, topSitesUsage]
  )
  const resetUsage = useCallback(() => {
    message(RESET_USAGE)
  }, [])

  // transforms
  const mappedTopSites = topSites.map(mapTopSites(topSitesUsage))
  const settingsInterface = useSettings()
  // TODO avoid mutation
  decorateSelectedSegment({ selectedSegment })

  return (
    <Provider
      value={{
        user,
        topSites: mappedTopSites,
        alarmEnabled,
        breathAmount,
        setBreathAmount,
        pageViews,
        showTopSites,
        siteTimes,
        appReady,
        breakTimers,
        selectedSegment,
        playerIsOpen,
        activityNumber,
        upsellModal,
        setAlarmEnabled: setAlarmEnabledProxy,
        setTopSites,
        updateSitesUsed,
        resetUsage,
        updateShowTopSites,
        setBreakTimer,
        setSelectedSegment,
        setPlayerIsOpen,
        setUpsellModal,
        updateBreakTimers,
        ...settingsInterface,
        ...tabInterface,
        ...endScreenInterface,
      }}
    >
      <IngressProvider>{props.children}</IngressProvider>
    </Provider>
  )
}

export const useExtension = () => useContext(context)
