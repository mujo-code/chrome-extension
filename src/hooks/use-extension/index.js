import { IngressProvider } from '@mujo/ingress'
import React, {
  useEffect,
  useState,
  useContext,
  useCallback,
} from 'react'
import { useTranslation } from 'react-i18next'
import {
  NEW_TAB_CONNECTION,
  RESET_USAGE,
  APP_READY_KEY,
  SCREEN_TIME_PERMISSIONS,
  TRANSLATION_FILE,
  CURRENT_TAB_KEY,
} from '../../constants'
import { toSiteInfo } from '../../lib/aggregation'
import { message } from '../../lib/extension'
import { usePermissions } from '../use-permissions'
import { useSubscription } from '../use-subscription'
import { useSettings } from './settings'
import { decorateSelectedSegment, mapTopSites } from './transforms'
import { useBreaktimerCallback } from './use-breaktimer-callback'
import { useDeeplink } from './use-deeplink'
import { useModel } from './use-model'
import { useTabs } from './use-tabs'
import { useTopsitesAPI } from './use-topsites-api'

const context = React.createContext()
const { Provider } = context

export const ExtensionProvider = props => {
  // state
  const { shouldRegisterApp } = props
  const { t } = useTranslation(TRANSLATION_FILE)
  const [appReady, setAppReady] = useState(false)
  const [playerIsOpen, setPlayerIsOpen] = useState(false)
  const [selectedSegment, setSelectedSegment] = useState(null)
  const [upsellModal, setUpsellModal] = useState(null)
  const tabInterface = useTabs(CURRENT_TAB_KEY)
  const {
    alarmEnabled,
    setAlarmEnabled,
    topSites,
    setTopSites,
    siteTimes,
    setSiteTimes,
    topSitesUsage,
    setTopSitesUsage,
    pageViews,
    updatePageViews,
    pending,
    showTopSites,
    updateShowTopSites,
    breakTimers,
    updateBreakTimers,
    activityNumber,
    breathAmount,
    setBreathAmount,
  } = useModel()
  const { user } = useSubscription()
  const {
    hasPermission,
    requestPermissions,
    removePermissions,
  } = usePermissions(SCREEN_TIME_PERMISSIONS)

  // effects
  useDeeplink({
    appReady,
    setPlayerIsOpen,
    setSelectedSegment,
  })
  useTopsitesAPI({ setTopSites })
  useEffect(() => {
    if (!pending && !appReady && shouldRegisterApp) {
      updatePageViews(pageViews + 1)
      message(NEW_TAB_CONNECTION)
      setAppReady(true)
      if (typeof window[APP_READY_KEY] === 'function') {
        window[APP_READY_KEY]()
      }
      window[APP_READY_KEY] = true
    }
  }, [
    pageViews,
    pending,
    setSiteTimes,
    setTopSites,
    updatePageViews,
    appReady,
    shouldRegisterApp,
  ])

  // callbacks
  const setAlarmEnabledProxy = useCallback(() => {
    const enabled = !alarmEnabled
    setAlarmEnabled(enabled)
  }, [alarmEnabled, setAlarmEnabled])
  const { setBreakTimer } = useBreaktimerCallback({
    breakTimers,
    setUpsellModal,
    updateBreakTimers,
    user,
  })
  const updateSitesUsed = useCallback(
    site => {
      setTopSitesUsage([...topSitesUsage, site])
    },
    [setTopSitesUsage, topSitesUsage]
  )
  const resetUsage = useCallback(() => {
    message(RESET_USAGE)
  }, [])

  // transforms
  const mappedTopSites = topSites.map(mapTopSites(topSitesUsage))
  const siteTimesAndTimers = toSiteInfo(siteTimes, breakTimers)
  const settings = useSettings({
    user,
    setUpsellModal,
    alarmEnabled,
    setAlarmEnabled,
    hasPermission,
    requestPermissions,
    removePermissions,
    breathAmount,
    setBreathAmount,
    t,
  })
  // TODO avoid mutation
  decorateSelectedSegment({ selectedSegment, siteTimesAndTimers })

  return (
    <Provider
      value={{
        topSites: mappedTopSites,
        alarmEnabled,
        breathAmount,
        pageViews,
        showTopSites,
        siteTimes,
        siteTimesAndTimers,
        appReady,
        breakTimers,
        selectedSegment,
        playerIsOpen,
        activityNumber,
        upsellModal,
        settings,
        screenTime: {
          hasPermission,
          requestPermissions,
          removePermissions,
        },
        setAlarmEnabled: setAlarmEnabledProxy,
        setTopSites,
        updateSitesUsed,
        resetUsage,
        updateShowTopSites,
        setBreakTimer,
        setSelectedSegment,
        setPlayerIsOpen,
        setUpsellModal,
        updateBreakTimers,
        ...tabInterface,
      }}
    >
      <IngressProvider>{props.children}</IngressProvider>
    </Provider>
  )
}

export const useExtension = () => useContext(context)
