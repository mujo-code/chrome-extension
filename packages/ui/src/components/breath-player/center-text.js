import React from 'react'
import { useSpring, animated as A } from 'react-spring'
import { Box } from '../box'
import { useTheme } from '../../hooks/use-theme'
import { HeaderL } from '../fonts'

export const CenterText = ({ children, show, down, ...otherProps }) => {
  const scaleOut = down ? 'scale(0.9)' : 'scale(1.1)'
  const { background } = useTheme()
  const style = useSpring({
    opacity: show ? 1 : 0,
    transform: show ? 'scale(1)' : scaleOut,
    config: { duration: show ? 1000 : 500 },
    delay: show ? 1000 : 0,
  })
  return (
    <Box
      layer="2"
      display="flex"
      justifyContent="center"
      alignItems="center"
      position="absolute"
      Component={A.div}
      color={background}
      style={style}
      css={{
        left: 0,
        right: 0,
        top: 'calc(50vh - 20px)',
        bottom: 'calc(50vh + 20px)',
        textAlign: 'center',
        height: '40px',
      }}
      {...otherProps}
    >
      <HeaderL marginTop="zero" marginBottom="zero">
        {children}
      </HeaderL>
    </Box>
  )
}
