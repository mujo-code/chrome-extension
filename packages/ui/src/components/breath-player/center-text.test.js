import React from 'react'
import { create } from 'react-test-renderer'
import { ColorThemeProvider } from '../../hooks/use-theme'
import { CenterText } from './center-text'

test('CenterText component matches snapshot', () => {
  const tree = create(
    <ColorThemeProvider value="dark">
      <CenterText>Foo</CenterText>
    </ColorThemeProvider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
