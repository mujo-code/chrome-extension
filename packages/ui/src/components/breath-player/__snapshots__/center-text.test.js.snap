// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CenterText component matches snapshot 1`] = `
<div
  className="css-r3g0g0"
  color="outerSpace"
  layer="2"
  position="absolute"
  style={
    Object {
      "opacity": 0,
      "transform": "scale(1.1)",
    }
  }
>
  <h2
    className="css-2tlyqf"
    color="color"
  >
    Foo
  </h2>
</div>
`;
